cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,556 @@
+ * IotMqttPublishInfo
+ ****************************************************************/
+
+IotMqttPublishInfo_t *allocate_IotMqttPublishInfo(IotMqttPublishInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
+  if ( pInfo == NULL ) return NULL;
+
+  pInfo->pTopicName = malloc_can_fail( pInfo->topicNameLength );
+  // assumption here is checked below in valid_
+  __CPROVER_assume( VALID_CBMC_SIZE( pInfo->payloadLength ) );
+  pInfo->pPayload = malloc_can_fail( pInfo->payloadLength );
+  return pInfo;
+}
+
+bool valid_IotMqttPublishInfo(IotMqttPublishInfo_t *pInfo)
+{
+  return
+    pInfo &&
+
+    VALID_STRING(pInfo->pTopicName, pInfo->topicNameLength) &&
+    VALID_CBMC_SIZE( pInfo->topicNameLength ) &&
+    pInfo->pTopicName != NULL &&
+
+    VALID_STRING(pInfo->pPayload, pInfo->payloadLength) &&
+    VALID_CBMC_SIZE( pInfo->payloadLength ) &&
+
+    pInfo->retryMs <= IOT_MQTT_RETRY_MS_CEILING &&
+    pInfo->retryMs > 0  &&
+
+    // BUG: try without this name length assumption
+    pInfo->topicNameLength < 0xFFFF &&
+    // BUG: try without this name length assumption
+    pInfo->payloadLength > 0;
+}
+
+/****************************************************************
+ * IotNetworkConnection
+ ****************************************************************/
+
+void *allocate_IotNetworkConnection()
+{
+  return allocate_opaque_type();
+}
+
+/****************************************************************
+ * IotNetworkInterface
+ ****************************************************************/
+
+IotNetworkInterface_t *allocate_IotNetworkInterface()
+{
+  return malloc_can_fail( sizeof( IotNetworkInterface_t ) );
+}
+
+bool valid_IotNetworkInterface(IotNetworkInterface_t *netif)
+{
+  return ( netif != NULL );
+}
+
+bool stubbed_IotNetworkInterface(IotNetworkInterface_t *netif)
+{
+  return
+    IS_STUBBED_NETWORKIF_CREATE(netif) &&
+    IS_STUBBED_NETWORKIF_CLOSE(netif) &&
+    IS_STUBBED_NETWORKIF_SEND(netif) &&
+    IS_STUBBED_NETWORKIF_RECEIVE(netif) &&
+    IS_STUBBED_NETWORKIF_SETRECEIVECALLBACK(netif) &&
+    IS_STUBBED_NETWORKIF_SETCLOSECALLBACK(netif) &&
+    IS_STUBBED_NETWORKIF_DESTROY(netif);
+}
+
+/****************************************************************
+ * IotNetworkInterface stubs
+ ****************************************************************/
+
+IotNetworkError_t IotNetworkInterfaceCreate( void * pConnectionInfo,
+					     void * pCredentialInfo,
+					     void * pConnection )
+{
+  // pCredentialInfo can be null
+  // create accepts NULL credentials when there is no TLS configuration.
+  __CPROVER_assert(pConnectionInfo != NULL,
+		   "IotNetworkInterfaceCreate pConnectionInfo is not NULL");
+  __CPROVER_assert(pConnection != NULL,
+		   "IotNetworkInterfaceCreate pConnection is not NULL");
+
+  // create the connection
+  *(void **)pConnection = allocate_IotNetworkConnection();
+
+  IotNetworkError_t error;
+  return error;
+}
+
+#define MAX_TRIES 2
```

Comment:
Should this be defined in the Makefile?

Reply:
I don't want to clutter the makefiles with this detail, but I've surrounded this with an ifndef/endif so that it can be easily modified from a makefile if we choose to do this in the future.